### DISCUSSION TITLE
IDP Solution Design

### DISCUSSION THREAD

**Discussion by mhutter** [zitadel/zitadel#8263] [Q&A] [answered]
Hi!

How would I implement the following scenario, and I am looking for pointers on how to implement this using ZITADEL.

## Scenario

TL;DR: I am hosting a bunch of applications on a platform, and want to provide a centralized IDP to access all of them. All applications support OAuth.

I want to provide a centralized IDP so that users don't have to sign up for every application separately. To achieve that, I'm hosting ZITADEL at `id.example.com`

There are two types of applications:

1. Internal - Things like monitoring and management systems (think: Grafana, ArgoCD, ...)
2. Public - Hosted public third party applications (think: Docspell, Miniflux, ...)

Similarly, there are two kinds of users:

1. Internal users - They must authenticate via ZITADEL (with higher enforced security standards), and should have access to both Internal and Public applications.
2. Public users - They may sign up themselves using username/password, or optionally via a social login provider. They may only have access to Public applications; access to those should be granted automatically for all users. No Federation or similar will happen here.

## Questions

My questions are:

- In general, is this doable with ZITADEL?
- How do I set up the Organizations, Projects and Apps?

From reading the docs, I'm unsure whether the correct way would be to just have 2 organizations:

1. Public - holds the public users and applications (also granted to users in Internal)
2. Internal - holds the internal users and applications

Or should the public applications be defined somewhere else?

**Comment by hifabienne** [comment, marked as answer]
Hi @mhutter 

Yes this is doable with ZITADEL.

I would set up 3 organizations.
1. Admin: with the ZITADEL projects and maybe an admin user, just to ensure you always have access and do not missconfigure anything there.
2. Your own organization: Here I would add your internal users and all the projects internal and external
The projects that should also be available to the other user can be granted to the 3rd organization
3. Public user org, this organization I would mark as default organization. The settings of the default organization are shown per default  when you redirect to the login.
You can configure the idps you need on this organization and allow registration 

After that you need to consider some thing, do you have specific roles on the project that you want to give the users? If yes you will have to add the roles to the project and also grant them the the other users. You can automatically add that with a custom action, (javascript snippet https://zitadel.com/docs/apis/actions/code-examples#add-grants-to-users)
You can either check the roles in your applications, or you can configure ZITADEL to check if the project  or an authorization is existing for the user who wants to authenticate: 
<img width="1093" alt="image" src="https://github.com/zitadel/zitadel/assets/38692350/d2059b3d-0acf-4c42-a487-da359c279f52">

**Reply by mhutter** [reply]
Thank you, this helped a lot!
